File: BinarySearchTree/src/main/java/Solution528.java

import common.TreeNode;

public class Solution528 {
    int sum = 0;

    public TreeNode convertBST(TreeNode root) {
        dfs(root);
        return root;

    }

    private void dfs(TreeNode root) {
        if (root != null) {
            dfs(root.right);
            sum = sum + root.val;
            root.val = sum;
            dfs(root.left);
        }

        return;

    }

    public static void main(String[] args) {
        Solution528 s = new Solution528();
        TreeNode root = new TreeNode(4);
        TreeNode left = new TreeNode(3);
        TreeNode right = new TreeNode(6);
        TreeNode rightLeft = new TreeNode(5);
        TreeNode rightRight = new TreeNode(7);

        right.left = rightLeft;
        right.right = rightRight;

        root.left = left;
        root.right = right;
        s.convertBST(root);

    }

}

File: BinarySearchTree/src/main/python/solution528.py
from typing import Optional


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def convertBST(self, root: Optional[TreeNode]) -> Optional[TreeNode]:
        if root == None:
            return
        ans = []

        def get_in_order(root):
            if root.right == None:
                ans.append(root.val)
                if root.left:
                    get_in_order(root.left)
                return

            if root.right:
                get_in_order(root.right)

            ans.append(root.val)

            if root.left:
                get_in_order(root.left)

        get_in_order(root)
        pre_sum = [0 for _ in range(len(ans) + 1)]
        for i in range(1, len(ans) + 1):
            pre_sum[i] = pre_sum[i - 1] + ans[i - 1]

        res_map = dict(zip(ans, pre_sum[1:]))

        def change_vale(root):
            if root == None:
                return

            root.val = res_map.get(root.val)

            if root.right:
                change_vale(root.right)
            if root.left:
                change_vale(root.left)

        change_vale(root)
        return root